src/v/transform/logging/event.cc

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "transform/logging/event.h"
+
+#include "json/json.h"
+
+using namespace std::chrono_literals;
+
+namespace transform::logging {
+
+using otel_xform_name_attr
+  = named_type<std::string_view, struct otel_xform_name_attr_tag>;
+
+using otel_node_id_attr
+  = named_type<model::node_id, struct otel_node_id_attr_tag>;
+
+struct otel_log_event {
+    otel_log_event() = delete;
+    explicit otel_log_event(std::string_view name, const event& ev)
+      : name(name)
+      , source_id(ev.source_id)
+      , ts(ev.ts.time_since_epoch() / 1ns)
+      , level(ev.level)
+      , message(ev.message) {}
+    otel_xform_name_attr name;
+    otel_node_id_attr source_id;
+    uint64_t ts;
+    ss::log_level level;
+    std::string_view message;
+};
+} // namespace transform::logging
+
+namespace json {
+
+namespace tlg = transform::logging;
+
+inline void rjson_serialize(
+  Writer<StringBuffer>& w, const tlg::otel_xform_name_attr& name) {
+    w.StartObject();
+    w.Key("key");
+    w.String("transform_name");
+    w.Key("value");
+    w.StartObject();
+    w.Key("stringValue");
+    rjson_serialize(w, name());
+    w.EndObject();
+    w.EndObject();
+}
+
+inline void
+rjson_serialize(Writer<StringBuffer>& w, const tlg::otel_node_id_attr& nid) {
+    w.StartObject();
+    w.Key("key");
+    w.String("node");
+    w.Key("value");
+    w.StartObject();
+    w.Key("intValue");
+    rjson_serialize(w, nid());
+    w.EndObject();
+    w.EndObject();
+}
+
+inline void
+rjson_serialize(Writer<StringBuffer>& w, const tlg::otel_log_event& ev) {
+    w.StartObject();
+    w.Key("body");
+    rjson_serialize(w, ev.message);
+    w.Key("timeUnixNano");
+    rjson_serialize(w, ev.ts);
+    w.Key("severityNumber");
+    w.Uint(tlg::log_level_to_severity(ev.level));
+    w.Key("attributes");
+    w.StartArray();
+    rjson_serialize(w, ev.name);
+    rjson_serialize(w, ev.source_id);
+    w.EndArray();
+    w.EndObject();
+}
+
+} // namespace json
+
+namespace transform::logging {
+
+uint32_t log_level_to_severity(ss::log_level lvl) {
+    static const std::unordered_map<ss::log_level, int> severity{
+      {ss::log_level::trace, 1},
+      {ss::log_level::debug, 5},
+      {ss::log_level::info, 9},
+      {ss::log_level::warn, 13},
+      {ss::log_level::error, 17},
+    };
+    return severity.at(lvl);
+}
+
+event::event(
+  model::node_id source,
+  clock_type::time_point ts,
+  ss::log_level level,
+  ss::sstring message)
+  : source_id(source)
+  , ts(ts)
+  , level(level)
+  , message(std::move(message)) {}
+
+void event::to_json(std::string_view name, iobuf& b) const {
+    ::json::StringBuffer buf;
```

Comment:
I believe we can write directly to an iobuf, code example: https://github.com/redpanda-data/redpanda/blob/6ba49d986ff09c7e12d43bd8193460f38654bc20/src/v/cluster/cloud_metadata/cluster_manifest.cc#L107-L116

Reply:
oh, neat

Reply:
done